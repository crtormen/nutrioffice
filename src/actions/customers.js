import database from '../firebase/firebase';

//Action generator ADD_CUSTOMER
export const addCustomer = (customer) => ({
    type: 'ADD_CUSTOMER',
    customer
});

//Pass a function instead of a object  
// export const startAddCustomer = (customerData = {}) => {
//     return (dispatch, getState) => { 
//         const uid = getState().auth.uid;
//         const { //default values
//             name = '', 
//             email = '',
//             phone = '', 
//             born = '',              
//             occupation = '', 
//             instagram = ''
//         } = customerData;
//         const customer = { name, email, phone, born, occupation, instagram };

//         return database.ref(`users/${uid}/customers`).push(customer).then((ref) => { 
//             dispatch(addCustomer({
//                 id: ref.key,
//                 ...customer
//             }));
//         });
//     };
// };

//Action generator for EDIT_CUSTOMER
export const editCustomer = (id, updates) => ({
    type: 'EDIT_CUSTOMER',
    id,
    updates
});

// export const startEditExpense = (id, updates) => {
//     return (dispatch, getState) => {
//         const uid = getState().auth.uid;
//         return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
//             dispatch(editExpense(id, updates));
//         })
//     };
// };

//Action generator for REMOVE_CUSTOMER
export const removeCustomer = ({ id } = {}) => ({
    type: 'REMOVE_CUSTOMER',
    id
});

// export const startRemoveExpense = ({ id } = {}) => {
//     return (dispatch, getState) => {
//         const uid = getState().auth.uid;
//         return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
//             dispatch(removeExpense({ id }));
//         });
//     };
// };


//Action generator for SET_CUSTOMERS
export const setCustomers = (customers) => ({
    type: 'SET_CUSTOMERS',
    customers
});

//Action Generator START_SET_EXPENSES
// export const startSetExpenses = () => {
//     return (dispatch, getState) => {
//         const uid = getState().auth.uid;
//         //returns a promises, so that when using startSetExpenses, we can use "then"
//         return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
//             let expenses = [];
//             snapshot.forEach((childSnapshot) => {
//                 expenses.push({
//                     id: childSnapshot.key,
//                     ...childSnapshot.val()
//                 })
//             });

//             dispatch(setExpenses(expenses));
//         });        
//     };
// };
