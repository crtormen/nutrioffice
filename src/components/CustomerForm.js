import React from 'react';


export default class CustomerForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            name: props.customer ? props.customer.name : '',
            email: props.customer ? props.customer.email : '',
            phone: props.customer ? props.customer.phone : '',
            born: props.customer ? props.customer.born : '',
            occupation: props.customer ? props.customer.occupation : '',
            instagram: props.customer ? props.customer.instagram : '',
           // calendarFocused: false,
            error: ''
        };
    }


    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.name || !this.state.email || !this.state.phone) {
            this.setState(() => ({ error: "VocÃª deve informar seu nome, email e celular."})); 
        } else {
            this.setState(() => ({ error: ""})); 
            // Runs parent method
            this.props.onSubmit({
                name: this.state.name,
                email: this.state.email,
                phone: this.state.phone,
                born: this.state.born,
                occupation: this.state.occupation,
                instagram: this.state.instagram                 
            });
        };

    }

    onNameChange = (e) => {
        const name = e.target.value;
        this.setState(() => ({ name }));
    };

    onEmailChange = (e) => {
        const email = e.target.value;
        this.setState(() => ({ email }));
    };

    onPhoneChange = (e) => {
        const phone = e.target.value;
        this.setState(() => ({ phone }));
    };

    onOccupationChange = (e) => {
        const occupation = e.target.value;
        this.setState(() => ({ occupation }));
    };

    onBornChange = (e) => {
        const born = e.target.value;
        this.setState(() => ({ born }));
    };

    onInstagramChange = (e) => {
        const instagram = e.target.value;
        this.setState(() => ({ instagram }));
    };

    render() {
        return (
            <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}>
                    <input 
                        type="text" 
                        placeholder="Nome completo" 
                        autoFocus 
                        value={this.state.name}
                        onChange={this.onNameChange}
                    />
                    <input 
                        type="text" 
                        placeholder="Data de nascimento" 
                        value={this.state.born}
                        onChange={this.onBornChange}
                    />
                    <input 
                        type="text" 
                        placeholder="Email"
                        value={this.state.email}
                        onChange={this.onEmailChange}
                    />
                    <input 
                        type="text" 
                        placeholder="Celular" 
                        value={this.state.phone}
                        onChange={this.onPhoneChange}
                    />
                    <input 
                        type="text" 
                        placeholder="ProfissÃ£o" 
                        value={this.state.occupation}
                        onChange={this.onOccupationChange}
                    />
                    <input 
                        type="text" 
                        placeholder="instagram" 
                        value={this.state.instagram}
                        onChange={this.onInstagramChange}
                    />
                <button>Salvar Dados</button>
            </form>
            </div>
        );
    }
}
