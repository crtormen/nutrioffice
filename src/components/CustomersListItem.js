import React from 'react';
import { Link } from 'react-router-dom';

const CustomersListItem = ({id, name, email}) => (
    <div>
        <h3><Link to={`edit/${id}`}>{name}</Link> - {email}</h3>    
    </div>
);

export default CustomersListItem;