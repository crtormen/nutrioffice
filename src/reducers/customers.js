const customerReducerDefaultState = [];

export default (state = customerReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_CUSTOMER':
            return [...state, action.customer];
        case 'EDIT_CUSTOMER':
            return state.map((customer) => {
                if (customer.id === action.id) {
                    return {
                        ...customer,
                        ...action.updates //create new object based ond the previous, with updates and concatenations
                    }
                }
                return customer; //do nothing if not found
            });
        case 'REMOVE_CUSTOMER':
            return state.filter(({ id }) => id !== action.id ); ////return all in a new array, except that one where filter not match 
        case 'SET_CUSTOMERS':
            return action.customers;
        default: 
            return state;
    };
};