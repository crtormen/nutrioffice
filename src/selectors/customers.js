export default (customers) => {
    return customers;
}



// Get visible expenses, filtering by text, period of time, and allows to sort by date and amount
// export default (expenses, { text, sortBy, startDate, endDate }) => { 
//     return expenses.filter((expense) => {
//         const startDateMatch = startDate ? startDate.isSameOrBefore(moment(expense.createdAt), 'day') : true;
//         const endDateMatch = endDate ? endDate.isSameOrAfter(moment(expense.createdAt), 'day') : true;
//         const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

//         return startDateMatch && endDateMatch && textMatch;
//     }).sort((a, b) => {  //highest first
//         if (sortBy === 'date')
//             return a.createdAt < b.createdAt ? 1 : -1;
//         if (sortBy === 'amount')
//             return a.amount < b.amount ? 1 : -1;

//     });
// };
