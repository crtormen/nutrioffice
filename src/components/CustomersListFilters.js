import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByName } from '../actions/filters';

export class CustomersListFilters extends React.Component {
    
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    render() {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange} 
                    placeholder="Busca..."
                />
            </div>
        );
    };
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByName: () => dispatch(sortByName())
});

const mapStateToProps = (state) => ({
    filters: state.filters
});

export default connect(mapStateToProps, mapDispatchToProps)(CustomersListFilters);