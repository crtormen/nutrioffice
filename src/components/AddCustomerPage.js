import React from 'react';
import { connect } from 'react-redux';
import uuid from 'uuid';
import { addCustomer } from '../actions/customers';
import CustomerForm from './CustomerForm';

export class AddCustomerPage extends React.Component {

    onSubmit = (customer) => {
        //TODO: props.startAddCustomer
        this.props.addCustomer({
            id: uuid(),
            ...customer
        });
        this.props.history.push('/dashboard');
    }

    render() {
        return (
            <div>
                <h2>Novo Paciente</h2>
                <CustomerForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    addCustomer: (customer) => dispatch(addCustomer(customer))
});

export default connect(undefined, mapDispatchToProps)(AddCustomerPage);
