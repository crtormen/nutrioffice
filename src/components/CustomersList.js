import React from 'react';
import { connect } from 'react-redux';
import selectCustomers from '../selectors/customers';
import CustomersListItem from './CustomersListItem';

export const CustomersList = (props) => (
    <div>
    <h2>Customers List</h2>
    {
        props.customers.length === 0 ? (
            <p>Nenhum paciente encontrado.</p>
        ) : (
            props.customers.map((customer) => {
                return (
                    <CustomersListItem 
                        key={customer.id}
                        {...customer}
                    />
                )
            })
        )
    }
    </div>
);

const mapStateToProps = (state, props) => ({
    customers: selectCustomers(state.customers)
});

//connect to store to receive and show always the most updated list of customers
export default connect(mapStateToProps)(CustomersList);