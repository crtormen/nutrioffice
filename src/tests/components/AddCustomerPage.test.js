import React from 'react';
import { shallow } from 'enzyme';
import { AddCustomerPage } from '../../components/AddCustomerPage';
import { customers } from '../fixtures/customers';

let wrapper, history, addCustomer;

beforeEach(() => {
    addCustomer = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddCustomerPage addCustomer={addCustomer} history={history}/>);
});

test("should render AddCustomerPage correctly", () => {
    expect(wrapper).toMatchSnapshot();
});

test("should handle onSubmit", () => {
    wrapper.find('CustomerForm').prop('onSubmit')(customers[1]);
    expect(history.push).lastCalledWith('/dashboard');
    expect(addCustomer).lastCalledWith(customers[1]);
});