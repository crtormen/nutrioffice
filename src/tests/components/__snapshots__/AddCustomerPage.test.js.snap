// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddCustomerPage correctly 1`] = `
<div>
  <h2>
    Novo Paciente
  </h2>
  <CustomerForm
    onSubmit={[Function]}
  />
</div>
`;
